ggshield/oauth.py

```diff
@@ -0,0 +1,240 @@
+        - retrieve the oauth authorization code from the request uri
+        - exchange the authorization code against an access token user gg's api
+        - save the access token
+        """
+        # enable redirection to http://localhost
+        os.environ["OAUTHLIB_INSECURE_TRANSPORT"] = str(True)
+
+        self._redirect_to_login()
+        authorization_code = self._wait_for_code()
+
+        self._claim_token(authorization_code)
+        token_data = self._validate_access_token()
+
+        self._save_token(token_data)
+
+        click.echo("Authentication was successful.")
+        self._redirect_to_authenticated()
+
+    def get_code(self, uri: str) -> Optional[str]:
+        """
+        extract the authorization from the incoming request uri
+        if no code can be extracted, return None
+        """
+        return self._oauth_client.parse_request_uri_response(uri).get("code")  # type: ignore
+
+    def _generate_pkce_pair(self) -> None:
+        """
+        Generate a code verifier (random string) and it's sha encoded version to be used
+        for the pkce checking process
+        """
+        self.code_verifier = token_urlsafe(96)[:128]
+        self.code_challenge = (
+            urlsafe_b64encode(sha256(self.code_verifier.encode()).digest())
+            .decode()
+            .rstrip("=")
+        )
+
+    def _redirect_to_login(self) -> None:
+        """
+        Open the user's browser to the GitGuardian ggshield authentication page
+        """
+        static_params = {
+            "auth_mode": "ggshield_login",
+            "utm_source": "cli",
+            "utm_medium": "login",
+            "utm_campaign": "ggshield",
+        }
+        request_uri = self._oauth_client.prepare_request_uri(
+            uri=urlparse.urljoin(self.dashboard_url, "auth/login"),
+            redirect_uri=self.redirect_uri,
+            scope=SCOPE,
+            code_challenge=self.code_challenge,
+            code_challenge_method="S256",
+            **static_params,
+        )
+        click.echo(
+            "Complete the login via GitGuardian client. Launching browser to:\n"
+            + str(request_uri)
+        )
+        webbrowser.open_new_tab(request_uri)
+
+    def _redirect_to_authenticated(self) -> None:
+        """
+        Open the authenticated page in the user's browser
+        """
+        webbrowser.open(urlparse.urljoin(self.dashboard_url, "authenticated"), new=0)
+
+    def _wait_for_code(self) -> str:
+        """
+        Open a local server on the predetermined port to receive the post-login redirection
+        including the authorazation code.
+        It uses a special request handler which is able to kill the server once the code
+        is received
+        return the authorization code or None if none could be fetched.
+        """
+        handler_wrapper = RequestHandlerWrapper(oauth_client=self)
+        server = HTTPServer(("", REDIRECT_PORT), handler_wrapper.request_handler)
```

Comment:
You're opening HTTPServer to "0.0.0.0", meaning that it will be exposed to all your interfaces. You should restrict to localhost.